Check Duplicates for Lazy and Provider

kshirsagarps opened the pull request:
We are checking for cases below, added unit tests and also verified with a small sample application.

1. Lazy injection in the superclass, duplicate regular injection in the subclass.
2. Regular injection in the superclass, duplicate Lazy injection in the subclass.
3. Lazy injection in the superclass and duplicate Lazy injection in the subclass.
4. Provider Injection in the superclass and duplicate provider injection in the subclass.
5. Lazy provider injection in the superclass and duplicate regular provider injection in the subclass.
6. Provider Injection in the superclass and duplicate lazy provider injection in the subclass.
7. Lazy Provider in the superclass and duplicate lazy provider injection in the subclass.

stephanenicolas commented on compiler/src/main/java/com/groupon/android/dichecks/checks/duplicate/DuplicateInjectionInHierarchyCheck.java:
```diff
@@ -126,6 +141,43 @@ private TypeElement findSuperClass(TypeElement typeElement) {
     return (TypeElement) typeUtils.asElement(typeElement.getSuperclass());
   }
 
+  private boolean isSubType(TypeMirror typeMirror, String typeName) {
```
I advise you to have more balanced signature (type mirror for both arguments), so that you deal with an equal level of abstraction.

stephanenicolas commented on compiler/src/main/java/com/groupon/android/dichecks/checks/duplicate/DuplicateInjectionInHierarchyCheck.java:
```diff
@@ -48,12 +55,20 @@ public DuplicateInjectionInHierarchyCheck(
   public void addInjectedElements(Set<? extends Element> injectedElements) {
     for (Element injectedElement : injectedElements) {
-      addInjectionDefinition(new InjectionDefinition(injectedElement), injectedElement);
+      if (isProvider(injectedElement)) {
```
or hasName("Lazy"), later

stephanenicolas commented on compiler/src/main/java/com/groupon/android/dichecks/checks/duplicate/DuplicateInjectionInHierarchyCheck.java:
```diff
@@ -126,6 +141,43 @@ private TypeElement findSuperClass(TypeElement typeElement) {
+  @Nullable
+  private TypeElement getKindParameter(Element element) {
+    final TypeMirror type = element.asType();
+    if (type.getKind() == TypeKind.DECLARED) {
```
I would flip the if (ALT + ENTER), so that edge cases are solved first.

stephanenicolas commented on compiler/src/main/java/com/groupon/android/dichecks/checks/duplicate/DuplicateInjectionInHierarchyCheck.java:
```diff
@@ -126,6 +141,43 @@ private TypeElement findSuperClass(TypeElement typeElement) {
+      final List<? extends TypeMirror> typeMirrors =
+          ((DeclaredType) element.asType()).getTypeArguments();
+      if (!typeMirrors.isEmpty()) {
+        return (TypeElement) typeUtils.asElement(typeUtils.erasure(typeMirrors.get(0)));
```
you should check the type arguments have a size of 1

stephanenicolas commented on compiler/src/test/java/com/groupon/android/dichecks/ProcessorTest.java:
```diff
@@ -104,6 +104,231 @@ public void compilationShoulNotFailIfDuplicateFoundAndWarningFlagSet() {
+                    "    @Inject Lazy<InjectedClass> aClass;",
+                    "}",
+                    "class D extends B {}",
+                    "class C extends A {}",
```
why C & D ?

stephanenicolas commented on compiler/src/test/java/com/groupon/android/dichecks/ProcessorTest.java:
```diff
@@ -104,6 +104,231 @@ public void compilationShoulNotFailIfDuplicateFoundAndWarningFlagSet() {
+                      "    @Inject InjectedClass aClass;",
+                      "}",
+                      "class D extends B {}",
+                      "class C extends A {}",
```
why C & D?

stephanenicolas commented on compiler/src/test/java/com/groupon/android/dichecks/ProcessorTest.java:
```diff
@@ -104,6 +104,231 @@ public void compilationShoulNotFailIfDuplicateFoundAndWarningFlagSet() {
+                    "}",
+                    "class D extends B {}",
+                    "class C extends A {}",
+                    "class InjectedClass {}",
```
same

stephanenicolas commented on compiler/src/test/java/com/groupon/android/dichecks/ProcessorTest.java:
```diff
@@ -104,6 +104,231 @@ public void compilationShoulNotFailIfDuplicateFoundAndWarningFlagSet() {
+                      "import javax.inject.Provider;",
+                      "public class A {",
+                      "    @Inject @Named(\"someNamedString\") String something;",
+                      "    @Inject Lazy<InjectedClass> aClass;",
```
you test too many things (apply to all tests). Moreover, fails to compile could check a specific error message.

code-twister commented on compiler/src/test/java/com/groupon/android/dichecks/ProcessorTest.java:
```diff
@@ -104,6 +104,231 @@ public void compilationShoulNotFailIfDuplicateFoundAndWarningFlagSet() {
+                    "    @Inject StringProvider stringProvider;",
+                    "}",
+                    "class B extends A {",
+                    "    @Inject StringProvider stringProvider;",
```
This case should be covered by the previous implementation. It's the same Class used in Both A and B. The provider case is a bit tricky. The situation I expect to cause us headaches is when you subclass Provider<SomeClass> and inject it somewhere, but use SomeClass directly in another place. If this is a valid use case we want to prevent you should add a test for that.

code-twister commented:
Looks good 👍  once you've addressed the comments we can take another look.

stephanenicolas commented on compiler/src/main/java/com/groupon/android/dichecks/checks/duplicate/DuplicateInjectionInHierarchyCheck.java:
```diff
@@ -126,6 +143,58 @@ private TypeElement findSuperClass(TypeElement typeElement) {
+        element.asType(), elementUtils.getTypeElement(Provider.class.getCanonicalName()).asType());
+  }
+
+  private boolean isLazy(Element element) {
```
nice. Can you just add some comment to explain why we have to do this (for dagger lazys). Do it when we call this method.

stephanenicolas commented on compiler/src/main/java/com/groupon/android/dichecks/checks/duplicate/DuplicateInjectionInHierarchyCheck.java:
```diff
@@ -126,6 +143,58 @@ private TypeElement findSuperClass(TypeElement typeElement) {
+   * Determine and return the type of the element.
+   *
+   * <ul>
+   *   <li>Return erasure type Foo for, Lazy&lt;Foo&gt;.
```
erasure type --> parameter type.
Type erasure is the technique that removes the type parameter of generics at compilation time.

stephanenicolas commented on compiler/src/main/java/com/groupon/android/dichecks/checks/duplicate/DuplicateInjectionInHierarchyCheck.java:
```diff
@@ -126,6 +143,58 @@ private TypeElement findSuperClass(TypeElement typeElement) {
+   *
+   * <ul>
+   *   <li>Return erasure type Foo for, Lazy&lt;Foo&gt;.
+   *   <li>Return FooProvider if no erasure type.
```
same

stephanenicolas replied:
This case bugs me a bit. It's quite different from the other one. It has a smell that we don't have a full control over what we are doing.

stephanenicolas commented on compiler/src/main/java/com/groupon/android/dichecks/checks/duplicate/DuplicateInjectionInHierarchyCheck.java:
```diff
@@ -126,6 +143,58 @@ private TypeElement findSuperClass(TypeElement typeElement) {
+  }
+
+  /**
+   * Determine and return the type of the element.
```
Return the type parameter of a class that extends {@code Lazy<T>} or {@code Provider<T>}.

stephanenicolas commented on compiler/src/test/java/com/groupon/android/dichecks/ProcessorTest.java:
```diff
@@ -42,19 +44,16 @@ public void testIntegration() {
 
   @Test
-  public void compilationShoulFailIfDuplicateFound() {
+  public void compilationShouldFailIfDuplicateInjectionFound() {
```
In this class, I don't understand why there are 2 things getting injected: on line 66 & 67.

stephanenicolas commented on compiler/src/test/java/com/groupon/android/dichecks/ProcessorTest.java:
```diff
@@ -104,6 +104,118 @@ public void compilationShoulNotFailIfDuplicateFoundAndWarningFlagSet() {
+                      "import javax.inject.Named;",
+                      "import toothpick.Lazy;",
+                      "public class A {",
+                      "    @Inject @Named(\"someNamedString\") String something;",
```
rename this.

stephanenicolas commented on compiler/src/test/java/com/groupon/android/dichecks/ProcessorTest.java:
```diff
@@ -104,6 +104,118 @@ public void compilationShoulNotFailIfDuplicateFoundAndWarningFlagSet() {
+            "toothpick.Lazy",
+            Joiner.on('\n')
+                .join(
+                    " package toothpick;",
```
remove the space before `package`

stephanenicolas commented on compiler/src/test/java/com/groupon/android/dichecks/ProcessorTest.java:
```diff
@@ -104,6 +104,118 @@ public void compilationShoulNotFailIfDuplicateFoundAndWarningFlagSet() {
+            "dagger.Lazy",
+            Joiner.on('\n')
+            .join(
+                " package dagger;",
```
same

stephanenicolas commented on compiler/src/test/java/com/groupon/android/dichecks/ProcessorTest.java:
```diff
@@ -104,6 +104,118 @@ public void compilationShoulNotFailIfDuplicateFoundAndWarningFlagSet() {
+                    "import javax.inject.Named;",
+                    "import dagger.Lazy;",
+                    "public class A {",
+                    "    @Inject @Named(\"someNamedString\") String something;",
```
remove this

stephanenicolas commented on compiler/src/test/java/com/groupon/android/dichecks/ProcessorTest.java:
```diff
@@ -104,6 +104,118 @@ public void compilationShoulNotFailIfDuplicateFoundAndWarningFlagSet() {
+                    "class B extends A {",
+                    "    @Inject StringProvider<String> stringProvider;",
+                    "}",
+                    "class StringProvider<String> implements Provider {}"));
```
It's funny: I don't know what this class declaration is. Usually it should be `StringProvider extends Provider<String>` and you would have to implement the method get... How does this work ?

code-twister replied:
I guess you could name your generic Type String, but I think this is a typo. Please double check that it works for the case @stephanenicolas mentioned in the comment above.

stephanenicolas commented:
We are close but I think there is something not very clear about type erasure.

Here is my concern and a few test cases:
```java
//case 1: handled properly
A: @Inject Foo foo;
B extends A: @Inject Lazy<Foo> fooLazy;

//case 2: not sure it is handled properly
A: @Inject Foo foo;
B extends A: @Inject FooProvider fooProvider;
FooProvider implements Provider<Foo>

//case 3: not sure it is handled properly
A: @Inject FooProvider fooProvider0;
B extends A: @Inject FooProvider fooProvider1;
FooProvider implements Provider<Foo>
```
As you can see when there is a `Provider/Lazy<Foo>` then there can be 2 different ways to have duplicates. One with Foo, one with the Provider itself. I am not sure we handle this well. 
Start by adding the tests to see if we pass them.

Also, you use used a declaration that seems buggy to me : `FooProvider<Foo> extends Provider`, I don't understand what we did but it looks valid. I am quite surprised. What is this form exactly ? Should we handle it ? Is it a synonym for `FooProvider extends Provider<Foo>` ?

stephanenicolas commented:
For me once all tests are in, minor comments are addressed, all looks good. No need to further review on my side. Wait for @code-twister review and 🐑 it.

code-twister commented on compiler/src/main/java/com/groupon/android/dichecks/checks/duplicate/DuplicateInjectionInHierarchyCheck.java:
```diff
@@ -48,12 +57,20 @@ public DuplicateInjectionInHierarchyCheck(
   public void addInjectedElements(Set<? extends Element> injectedElements) {
     for (Element injectedElement : injectedElements) {
-      addInjectionDefinition(new InjectionDefinition(injectedElement), injectedElement);
+      if (isProvider(injectedElement) || isLazy(injectedElement)) {
```
To make this block of code easier to read we can extract a method that returns an InjectionDefinition, since in all cases we end up calling addInjectionDefinition().

code-twister commented on compiler/src/main/java/com/groupon/android/dichecks/checks/duplicate/DuplicateInjectionInHierarchyCheck.java:
```diff
@@ -126,6 +143,58 @@ private TypeElement findSuperClass(TypeElement typeElement) {
+    final TypeMirror type = element.asType();
+    if (type.getKind() == TypeKind.DECLARED) {
+      final Name className = ((DeclaredType) type).asElement().getSimpleName();
+      return LAZY_CLASS_NAME.equals(className.toString());
```
I like the Yoda condition here ;)

kshirsagarps commented:
@stephanenicolas and @code-twister I addressed all the comments. As per my conversation with @stephanenicolas we wanted to add suppression mechanism. Should we add it before we release the new version?

code-twister commented:
already started working on suppress mechanism, @kshirsagarps would you like to be involved?

kshirsagarps commented:
@code-twister yep, I would like to be part of it 👍
